shop/shop/src/main/java/ro/msg/learning/shop/services/strategies/MostAbundantStrategy.java

```diff
@@ -0,0 +1,54 @@
+package ro.msg.learning.shop.services.strategies;
+
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import ro.msg.learning.shop.entities.OrderDetail;
+import ro.msg.learning.shop.entities.Stock;
+import ro.msg.learning.shop.exceptions.MissingStockException;
+import ro.msg.learning.shop.services.StockService;
+import ro.msg.learning.shop.services.interfaces.IStrategy;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+@RequiredArgsConstructor
+public class MostAbundantStrategy implements IStrategy {
+    private final StockService stockService;
+
+    @Override
+    public List<Stock> createOrder(List<OrderDetail> products) {
+        if (null == products) {
+            throw new MissingStockException("Product list is empty");
+        }
+        List<Stock> stockList = stockService.findAll();
+        List<Stock> result = new ArrayList<>();
+        for (OrderDetail product : products) {
+            List<Stock> stocks = stockList.stream().filter(e -> e.getProduct().equals(product.getProduct()))
+                    .sorted(Comparator.comparingInt(Stock::getQuantity)).collect(Collectors.toList());
```

Comment:
Here you read the whole stock list from the DB to memory, filter it to include only the stocks for a particular product and then sort it to obtain the maximum. This is a little bit inefficient (memory + time). 

We could obtain the information we need using directly a repository query (it's more efficient to load from memory only want we actually need instead of loading everything and manipulating the data afterwards). 

Reference: [Working with Spring Data Repositories](https://docs.spring.io/spring-data/data-commons/docs/1.6.1.RELEASE/reference/html/repositories.html)

You need a custom query: something like `findTopByProductIdOrderByQuantityDesc`